Google.Api.Gax.Grpc/HeaderParameterExtractor.cs

```diff
@@ -0,0 +1,68 @@
+﻿/*
+ * Copyright 2021 Google Inc. All Rights Reserved.
+ * Use of this source code is governed by a BSD-style
+ * license that can be found in the LICENSE file or at
+ * https://developers.google.com/open-source/licenses/bsd
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+namespace Google.Api.Gax.Grpc
+{
+    /// <summary>
+    /// 
+    /// </summary>
+    public sealed class HeaderParameterExtractor<TRequest>
```

Comment:
I'd suggest making this immutable: pass all the regexes and all the selectors into the constructor in one go (possibly having an overload for the single-pattern version if that simplifies some generation situations). We can then create a `List(Regex, Func<TRequest, string>))` which has already been reversed. That feels better than doing the zipping and reversing on every request. (And immutability generally feels nicer...)

Reply:
👍

Reply:
Glad to to see it's now immutable - but we're still doing a lot of work on a per request basis. It might be easier for me to suggest a refactoring after merging though.